packages/crypto/src/transactions/types/transaction.ts

```diff
@@ -97,4 +98,45 @@ export abstract class Transaction implements ITransaction {
     public hasVendorField(): boolean {
         return false;
     }
+
+    /**
+     * Calculate the expiration height of a transaction.
+     * An expiration height H means that the transaction cannot be included in block at height
+     * H or any higher block.
+     * If the user did not specify an expiration height when creating the transaction then
+     * we calculate one from the timestamp of the transaction creation and the configured
+     * maximum transaction age.
+     * @return number expiration height or null if the transaction does not expire
+     */
+    public calculateExpiration(
+        context: {
+            blockTime: number;
+            currentHeight: number;
+            now: number;
+        },
+    ): number {
+        const data: ITransactionData = this.data;
+        // We ignore data.expiration in v1 transactions because it is not signed
+        // by the transaction creator.
+        // TODO: check if ok
+        if (data.version >= 2) {
+            // tslint:disable-next-line:no-null-keyword
+            return data.expiration || null;
+        }
+
+        const maxTransactionAge = app.resolveOptions("transaction-pool").maxTransactionAge;
```

Comment:
The crypto package is decoupled from core, it has no knowledge of the container and isn't allowed to have any.

It also shouldn't really be the task of the transaction itself to calculate the expiration but that is an issue for another day.